src/lib.rs

```diff
@@ -1,110 +1,451 @@
+//
+// The special handling of apostrophes is avoided in the
+// [`to_roff`](Roff::to_roff) method, but it's used in the
+// [`render`](Roff::render) and [`to_writer`](Roff::to_writer)
+// methods.
+//
+// See: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=507673#65
+const APOSTROPHE_PREABMLE: &str = r#".ie \n(.g .ds Aq \(aq
+.el .ds Aq '
+"#;
+
+// Use the apostrophe string variable.
+const APOSTROPHE: &str = r"\*(Aq";
+
+#[derive(Eq, PartialEq)]
+enum Apostrophes {
+    Handle,
+    DontHandle,
+}
+
+/// A ROFF document, consisting of lines.
+///
+/// Lines are either control lines (requests that are built in, or
+/// invocations of macros), or text lines.
+///
+/// # Example
+///
+/// ```
+/// # use roff::*;
+/// let doc = Builder::default()
+///     .control("TH", ["FOO", "1"])
+///     .control("SH", ["NAME"])
+///     .text([roman("foo - do a foo thing")])
+///     .build()
+///     .render();
+/// assert!(doc.ends_with(".TH FOO 1\n.SH NAME\nfoo \\- do a foo thing\n"));
+/// ```
+#[derive(Debug, PartialEq, Eq, Default)]
 pub struct Roff {
-    title: String,
-    section: i8,
-    content: Vec<Section>,
+    lines: Vec<Line>,
 }
 
 impl Roff {
-    pub fn new(title: &str, section: i8) -> Self {
-        Roff {
-            title: title.into(),
-            section,
-            content: Vec::new(),
-        }
+    /// Append a control line.
+    ///
+    /// The line consist of the name of a built-in command or macro,
+    /// and some number of arguments. Arguments that contain spaces
+    /// will be enclosed on double quotation marks.
+    pub fn control<'a>(&mut self, name: &'a str, args: impl Into<Vec<&'a str>>) {
+        self.lines.push(Line::control(
+            name.to_string(),
+            args.into().iter().map(|s| s.to_string()).collect(),
+        ));
     }
 
-    pub fn section<'a, C, I>(mut self, title: &str, content: I) -> Self
-    where
-        I: IntoIterator<Item = &'a C>,
-        C: Troffable + 'a,
-    {
-        let title = title.into();
-        let content = content.into_iter().map(|x| x.render()).collect();
-
-        self.content.push(Section { title, content });
-        self
+    /// Append a text line.
+    ///
+    /// The line will be rendered in a way that ensures it can't be
+    /// interpreted as a control line. The caller does not need to
+    /// ensure, for example, that the line doesn't start with a
+    /// period ("`.`") or an apostrophe ("`'`").
+    pub fn text(&mut self, inlines: impl Into<Vec<Inline>>) {
+        self.lines.push(Line::text(inlines.into()));
     }
-}
 
-impl Troffable for Roff {
-    fn render(&self) -> String {
-        let mut res = String::new();
+    /// Render as ROFF source text that can be fed to a ROFF implementation.
+    pub fn render(&self) -> String {
+        let mut buf = vec![];
+        self.to_writer(&mut buf).unwrap(); // writing to a Vec always works
+        String::from_utf8_lossy(&buf).into_owned()
```

Comment:
nit: since we guarantee all inputs are UTF-8, we can guarantee the output is UTF-8 and we can make this clearer by doing `std::str::from_utf8(&but).expect("output is utf8 if all input is utf8 and our API guarentees that").to_owned()`